src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,313 @@
+            Contract.ThrowIfNull(oldSemanticTokens.Data);
+            Contract.ThrowIfNull(newSemanticTokens.Data);
+
+            var edits = new SemanticTokensEdits
+            {
+                Edits = ComputeSemanticTokensEdits(oldSemanticTokens.Data, newSemanticTokens.Data),
+                ResultId = resultId
+            };
+
+            await _tokensCache.UpdateCacheAsync(request.TextDocument.Uri, newSemanticTokens, cancellationToken).ConfigureAwait(false);
+            return edits;
+        }
+
+        /// <summary>
+        /// Compares two sets of SemanticTokens and returns the edits between them.
+        /// </summary>
+        private static LSP.SemanticTokensEdit[] ComputeSemanticTokensEdits(
+            int[] oldSemanticTokens,
+            int[] newSemanticTokens)
+        {
+            // We use Roslyn's version of the Myers' Diff Algorithm to compute the minimal edits
+            // between the old and new tokens.
+            // Edits are computed by token (i.e. in sets of five integers), so if one value in the token
+            // is changed, the entire token is replaced. We do this instead of directly comparing each
+            // value in the token individually so that we can potentially save on computation costs, since
+            // we can return early if we find that one value in the token doesn't match. However, there
+            // are trade-offs since our insertions/deletions are usually larger.
+
+            // Turning arrays into tuples of five ints, each representing one token
+            var oldGroupedSemanticTokens = ConvertToGroupedSemanticTokens(oldSemanticTokens);
+            var newGroupedSemanticTokens = ConvertToGroupedSemanticTokens(newSemanticTokens);
+
+            var edits = LongestCommonSemanticTokensSubsequence.GetEdits(oldGroupedSemanticTokens, newGroupedSemanticTokens);
+
+            return ConvertToSemanticTokenEdits(newGroupedSemanticTokens, edits);
+        }
+
+        private static SemanticTokensEdit[] ConvertToSemanticTokenEdits(SemanticToken[] newGroupedSemanticTokens, IEnumerable<SequenceEdit> edits)
+        {
+            // Our goal is to minimize the number of edits we return to LSP. It's possible an index
+            // may have both an insertion and deletion, in which case we can combine the two into a
+            // single update. We use the dictionary below to keep track of whether an index contains
+            // an insertion, deletion, or both.
+            var indexToEditKinds = new Dictionary<int, EditKind>();
+
+            foreach (var edit in edits)
+            {
+                // We only care about EditKind.Insert and EditKind.Delete, since they encompass all
+                // changes to the document. All other EditKinds are ignored.
+                // However, we do use EditKind.Update to represent that an index contains both an
+                // insertion and deletion and thus can be combined.
```

Comment:
```suggestion
                // However, we do use EditKind.Update to represent that an index contains both an
                // insertion and deletion and thus can be combined.
```

remove.  or update to say you're specifically ignoring this one, and why.

Reply:
hrmm.. i'm probably wrong here.

Reply:
I realized the way I wrote out the logic was super confusing since I tried to re-use the existing EditKind enum to fulfill a different purpose. I pushed changes that use a new SemanticTokenEditKind enum instead that will hopefully make things clearer.